In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/the-movies-dataset/ratings.csv
/kaggle/input/the-movies-dataset/links_small.csv
/kaggle/input/the-movies-dataset/credits.csv
/kaggle/input/the-movies-dataset/keywords.csv
/kaggle/input/the-movies-dataset/movies_metadata.csv
/kaggle/input/the-movies-dataset/ratings_small.csv
/kaggle/input/the-movies-dataset/links.csv


# Movie Recommendation System

## Importing Necessary Libraries

In [2]:
import pandas as pd
pd.set_option("max_columns", 100)
import re

from sklearn.metrics.pairwise import linear_kernel

from nltk.corpus import stopwords
from nltk.stem import SnowballStemmer

import warnings
warnings.filterwarnings("ignore")

## Loading the Dataset

In [3]:
metadata = pd.read_csv("/kaggle/input/the-movies-dataset/movies_metadata.csv")

In [4]:
metadata.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview popularity  \
0  Led by Woody, Andy's toys live happily in his ...  21.946943   
1  When siblings Judy and Peter discover an encha...  17.015539   
2  A family wedding reignites the ancient feud be...    11.7129   
3  Cheated on, mistreated and stepped on, the wom...   3.859495   
4  Just when George Banks has recovered from his ...   8.387519   

                        poster_path  \
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   
3  /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg   
4  /e64sOI48hQXyru7naBFyssKFxVd.jpg   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3  [{'name': 'Twentieth Century Fox Film Corporat...   
4  [{'name': 'Sandollar Productions', 'id': 5842}...   

                                production_countries release_date  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   
2  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
3  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
4  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-02-10   

       revenue  runtime                                   spoken_languages  \
0  373554033.0     81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0    104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0    101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0    127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0    106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video  vote_average  vote_count  
0                    Toy Story  False           7.7      5415.0  
1                      Jumanji  False           6.9      2413.0  
2             Grumpier Old Men  False           6.5        92.0  
3            Waiting to Exh

## Popularity Based Recommendation

The formula for calculating the Top Rated 250 Titles gives a true Bayesian estimate:
weighted rating (WR) = (v ÷ (v+m)) × R + (m ÷ (v+m)) × C where:

- R = average for the movie (mean) = (Rating)
- v = number of votes for the movie = (votes)
- m = minimum votes required to be listed in the Top 250
- C = the mean vote across the whole report (currently 7.0)

#### The mean of the mean of all movies

In [5]:
C = metadata['vote_average'].mean()

#### Calculate the minimum number of votes required to be in the chart, m

- We will use cutoff m as the 90th percentile. In other words, for a movie to be featured in the charts, it must have more votes than at least 90% of the movies on the list. (In top of 10%)

In [6]:
m = metadata['vote_count'].quantile(0.90)

#### We copied the portion of the dataframe that we need and filtered it

In [7]:
movies = metadata.copy().loc[metadata['vote_count'] >= m]

In [8]:
movies.shape

(4555, 24)

#### Function that computes the weighted rating of each movie by IMDB formula

In [9]:
def weighted_rating(x, m = m, C = C):
    v = x['vote_count']
    R = x['vote_average']
    return (v / (v + m) * R) + (m / (m + v) * C)

#### We created new column has a name "Overall Score" and calculated its value with weighted_rating() function


In [10]:
movies['Overall Score'] = movies.apply(weighted_rating, axis = 1)

#### We sorted movies based on score calculated above

In [11]:
movies.sort_values(by = 'Overall Score', ascending = False, inplace = True)
movies[['title', 'vote_count', 'vote_average', 'Overall Score']].head(20)

title  vote_count  vote_average  \
314           The Shawshank Redemption      8358.0           8.5   
834                      The Godfather      6024.0           8.5   
10309      Dilwale Dulhania Le Jayenge       661.0           9.1   
12481                  The Dark Knight     12269.0           8.3   
2843                        Fight Club      9678.0           8.3   
292                       Pulp Fiction      8670.0           8.3   
522                   Schindler's List      4436.0           8.3   
23673                         Whiplash      4376.0           8.3   
5481                     Spirited Away      3968.0           8.3   
2211                 Life Is Beautiful      3643.0           8.3   
1178            The Godfather: Part II      3418.0           8.3   
1152   One Flew Over the Cuckoo's Nest      3001.0           8.3   
351                       Forrest Gump      8147.0           8.2   
1154           The Empire Strikes Back      5998.0           8.2   
1176                            Psycho      2405.0           8.3   
18465                 The Intouchables      5410.0           8.2   
40251                       Your Name.      1030.0           8.5   
289             Leon: The Professional      4293.0           8.2   
3030                    The Green Mile      4166.0           8.2   
1170                        GoodFellas      3211.0           8.2   

       Overall Score  
314         8.445869  
834         8.425439  
10309       8.421453  
12481       8.265477  
2843        8.256385  
292         8.251406  
522         8.206639  
23673       8.205404  
5481        8.196055  
2211        8.187171  
1178        8.180076  
1152        8.164256  
351         8.150272  
1154        8.132919  
1176        8.132715  
18465       8.125837  
40251       8.112532  
289         8.107234  
3030        8.104511  
1170        8.077459

#### That is all for popularity based recommender

## Content Based Recommendation

In [12]:
movies = metadata.copy()

In [13]:
movies

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
45461  False                                                NaN         0   
45462  False                                                NaN         0   
45463  False                                                NaN         0   
45464  False                                                NaN         0   
45465  False                                                NaN         0   

                                                  genres  \
0      [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1      [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2      [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3      [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                         [{'id': 35, 'name': 'Comedy'}]   
...                                                  ...   
45461  [{'id': 18, 'name': 'Drama'}, {'id': 10751, 'n...   
45462                      [{'id': 18, 'name': 'Drama'}]   
45463  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...   
45464                                                 []   
45465                                                 []   

                                   homepage      id    imdb_id  \
0      http://toystory.disney.com/toy-story     862  tt0114709   
1                                       NaN    8844  tt0113497   
2                                       NaN   15602  tt0113228   
3                                       NaN   31357  tt0114885   
4                                       NaN   11862  tt0113041   
...                                     ...     ...        ...   
45461  http://www.imdb.com/title/tt6209470/  439050  tt6209470   
45462                                   NaN  111109  tt2028550   
45463                                   NaN   67758  tt0303758   
45464                                   NaN  227506  tt0008536   
45465                                   NaN  461257  tt6980792   

      original_language               original_title  \
0                    en                    Toy Story   
1                    en                      Jumanji   
2                    en             Grumpier Old Men   
3                    en            Waiting to Exhale   
4                    en  Father of the Bride Part II   
...                 ...                          ...   
45461                fa                      رگ خواب   
45462                tl          Siglo ng Pagluluwal   
45463                en                     Betrayal   
45464                en          Satana likuyushchiy   
45465                en                     Queerama   

                                                overview popularity  \
0      Led by Woody, Andy's toys live happily in his ...  21.946943   
1      When siblings Judy and Peter discover an encha...  17.015539   
2      A family wedding reignites the ancient feud be...    11.7129   
3      Cheated on, mistreated and stepped on, the wom...   3.859495   
4      Just when George Banks has recovered from his ...   8.387519   
...                                                  ...        ...   
45461        Rising and falling between a man and woman.   0.072051   
45462  An artist struggles to finish his work while a...   0.178241   
45463  When one of her hits goes wrong, a professiona...   0.903007   
45464  In a small town live two brothers, one a minis...   0.003503   
45465  50 years after decriminalisation of homosexual...   0.163015   

                            poster_path  \

#### We turned string type list to list

In [14]:
from ast import literal_eval

movies["genres"] = movies["genres"].apply(literal_eval)

In [15]:
movies["genres"][0]

[{'id': 16, 'name': 'Animation'},
 {'id': 35, 'name': 'Comedy'},
 {'id': 10751, 'name': 'Family'}]

#### Genres extractor function

In [16]:
def genres_extractor(my_list):
    gen_list = []
    for i in my_list:
        gen_list.append(i["name"])
    
    return gen_list

#### We extracted the genres

In [17]:
movies["gen_list"] = movies["genres"].apply(genres_extractor)

In [18]:
movies.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview popularity  \
0  Led by Woody, Andy's toys live happily in his ...  21.946943   
1  When siblings Judy and Peter discover an encha...  17.015539   
2  A family wedding reignites the ancient feud be...    11.7129   
3  Cheated on, mistreated and stepped on, the wom...   3.859495   
4  Just when George Banks has recovered from his ...   8.387519   

                        poster_path  \
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   
3  /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg   
4  /e64sOI48hQXyru7naBFyssKFxVd.jpg   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3  [{'name': 'Twentieth Century Fox Film Corporat...   
4  [{'name': 'Sandollar Productions', 'id': 5842}...   

                                production_countries release_date  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   
2  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
3  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
4  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-02-10   

       revenue  runtime                                   spoken_languages  \
0  373554033.0     81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0    104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0    101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0    127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0    106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video  vote_average  vote_count  \
0                    Toy Story  False           7.7      5415.0   
1                      Jumanji  False           6.9      2413.0   
2             Grumpier Old Men  False           6.5        92.0   
3            Waiting to

#### We turned list to string

In [19]:
movies["gen_list"] = movies["gen_list"].apply(lambda x: " ".join(x)) 

#### We add gen_list to overview we will use it after

In [20]:
movies.overview = movies.gen_list + " " + movies.overview

In [21]:
movies.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview popularity  \
0  Animation Comedy Family Led by Woody, Andy's t...  21.946943   
1  Adventure Fantasy Family When siblings Judy an...  17.015539   
2  Romance Comedy A family wedding reignites the ...    11.7129   
3  Comedy Drama Romance Cheated on, mistreated an...   3.859495   
4  Comedy Just when George Banks has recovered fr...   8.387519   

                        poster_path  \
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   
3  /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg   
4  /e64sOI48hQXyru7naBFyssKFxVd.jpg   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3  [{'name': 'Twentieth Century Fox Film Corporat...   
4  [{'name': 'Sandollar Productions', 'id': 5842}...   

                                production_countries release_date  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   
2  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
3  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
4  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-02-10   

       revenue  runtime                                   spoken_languages  \
0  373554033.0     81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0    104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0    101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0    127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0    106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video  vote_average  vote_count  \
0                    Toy Story  False           7.7      5415.0   
1                      Jumanji  False           6.9      2413.0   
2             Grumpier Old Men  False           6.5        92.0   
3            Waiting to

#### We dropped null values

In [22]:
movies["overview"].dropna(inplace = True)

#### Preprocessing

#### Eliminating numbers, html tags from overview, shortly we are cleaning the overview text

In [23]:
stop_words = stopwords.words("english")
stemmer = SnowballStemmer("english")

cleaning = "@\S+|https?:\S+|http?:\S|[^A-Za-z]+"

def preprocess(text, stem=False):
    # Remove link,user and special characters
    text = re.sub(cleaning, ' ', str(text).lower()).strip()
    tokens = []
    for token in text.split():
        if token not in stop_words:
            if stem:
                tokens.append(stemmer.stem(token))
            else:
                tokens.append(token)
    return " ".join(tokens)

#### Applying cleaning function

In [24]:
movies.overview = movies.overview.apply(lambda x: preprocess(x, stem = False))

In [25]:
movies

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
45461  False                                                NaN         0   
45462  False                                                NaN         0   
45463  False                                                NaN         0   
45464  False                                                NaN         0   
45465  False                                                NaN         0   

                                                  genres  \
0      [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1      [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2      [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3      [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                         [{'id': 35, 'name': 'Comedy'}]   
...                                                  ...   
45461  [{'id': 18, 'name': 'Drama'}, {'id': 10751, 'n...   
45462                      [{'id': 18, 'name': 'Drama'}]   
45463  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...   
45464                                                 []   
45465                                                 []   

                                   homepage      id    imdb_id  \
0      http://toystory.disney.com/toy-story     862  tt0114709   
1                                       NaN    8844  tt0113497   
2                                       NaN   15602  tt0113228   
3                                       NaN   31357  tt0114885   
4                                       NaN   11862  tt0113041   
...                                     ...     ...        ...   
45461  http://www.imdb.com/title/tt6209470/  439050  tt6209470   
45462                                   NaN  111109  tt2028550   
45463                                   NaN   67758  tt0303758   
45464                                   NaN  227506  tt0008536   
45465                                   NaN  461257  tt6980792   

      original_language               original_title  \
0                    en                    Toy Story   
1                    en                      Jumanji   
2                    en             Grumpier Old Men   
3                    en            Waiting to Exhale   
4                    en  Father of the Bride Part II   
...                 ...                          ...   
45461                fa                      رگ خواب   
45462                tl          Siglo ng Pagluluwal   
45463                en                     Betrayal   
45464                en          Satana likuyushchiy   
45465                en                     Queerama   

                                                overview popularity  \
0      animation comedy family led woody andy toys li...  21.946943   
1      adventure fantasy family siblings judy peter d...  17.015539   
2      romance comedy family wedding reignites ancien...    11.7129   
3      comedy drama romance cheated mistreated steppe...   3.859495   
4      comedy george banks recovered daughter wedding...   8.387519   
...                                                  ...        ...   
45461              drama family rising falling man woman   0.072051   
45462  drama artist struggles finish work storyline c...   0.178241   
45463  action drama thriller one hits goes wrong prof...   0.903007   
45464  small town live two brothers one minister one ...   0.003503   
45465  years decriminalisation homosexuality uk direc...   0.163015   

                            poster_path  \

#### We will vectorize the all words in overview

In [26]:
from sklearn.feature_extraction.text import TfidfVectorizer

vect = TfidfVectorizer()

dtm = vect.fit_transform(movies["overview"])

In [27]:
dtm.shape

(45466, 73535)

#### We see that over 73535 different words were used to describe the 45466 movies in our dataset

#### We computed the cosine similarity matrix


In [28]:
dtm_matrix = dtm[:30000] # 30000 for prevent memory allocation problem
cosine_sim = linear_kernel(dtm_matrix, dtm_matrix)

In [29]:
cosine_sim.shape

(30000, 30000)

In [30]:
cosine_sim[1]

array([0.01878249, 1.        , 0.05101365, ..., 0.        , 0.        ,
       0.02264548])

#### Each movie will be a 1x30000 column vector where each column will be a similarity score with each movie.

#### We need to give their titles

In [31]:
indices = pd.Series(index = metadata['title'], data = metadata.index).drop_duplicates()
indices

title
Toy Story                          0
Jumanji                            1
Grumpier Old Men                   2
Waiting to Exhale                  3
Father of the Bride Part II        4
                               ...  
Subdue                         45461
Century of Birthing            45462
Betrayal                       45463
Satan Triumphant               45464
Queerama                       45465
Length: 45466, dtype: int64

#### Function that takes in movie title as input and outputs most similar movies

In [32]:
def get_recommendations(title, cosine_sim = cosine_sim):
    # Get the index of the movie that matches the title
    idx = indices[title]
    
    # Get the pairwsie similarity scores of all movies with that movie
    sim_scores = list(enumerate(cosine_sim[idx]))
    
    # Sort the movies based on the similarity scores
    sim_scores = sorted(sim_scores, key = lambda x : x[1], reverse = True)
    
    # Get the scores of the 10 most similar movies, we stared with 1 because index 0 will give the same name of the input title 
    sim_scores = sim_scores[1:11]
    
    # Get the movie indices
    movie_indices = [i[0] for i in sim_scores]
    
    # Return the top 10 most similar movies
    return metadata['title'][movie_indices]

#### Let's get recommendations

In [33]:
get_recommendations('The Dark Knight Rises')

12481                                      The Dark Knight
150                                         Batman Forever
1328                                        Batman Returns
585                                                 Batman
15511                           Batman: Under the Red Hood
21194    Batman Unmasked: The Psychology of the Dark Kn...
9230                    Batman Beyond: Return of the Joker
19792              Batman: The Dark Knight Returns, Part 1
18035                                     Batman: Year One
3095                          Batman: Mask of the Phantasm
Name: title, dtype: object

In [34]:
get_recommendations('The Godfather')

1178      The Godfather: Part II
1914     The Godfather: Part III
13346                  Game Over
8653                Violent City
23126                 Blood Ties
10821                   Election
11297           Household Saints
13177               I Am the Law
7515          The Valachi Papers
6711                    Mobsters
Name: title, dtype: object

#### We can also vectorize credits information such as director and cast names.

## User Based Recommendation Collaborative Filtering

In [35]:
#ratings = pd.read_csv('../input/the-movies-dataset/ratings_small.csv')
ratings = pd.read_csv('/kaggle/input/the-movies-dataset/ratings.csv')

ratings.drop("timestamp", axis = 1, inplace = True)

In [36]:
ratings

userId  movieId  rating
0              1      110     1.0
1              1      147     4.5
2              1      858     5.0
3              1     1221     5.0
4              1     1246     5.0
...          ...      ...     ...
26024284  270896    58559     5.0
26024285  270896    60069     5.0
26024286  270896    63082     4.5
26024287  270896    64957     4.5
26024288  270896    71878     2.0

[26024289 rows x 3 columns]

#### There are 270896 user, it is so much, our memory is not enough for this. So we take 15000 of them

In [37]:
ratings = ratings[ratings["userId"] < 15000]

#### We do some arrangements for movies dataframe

In [38]:
movies = metadata[["id", "title"]]
movies.columns = ["movieId", "title"]
movies = movies[~movies["movieId"].str.contains("-")]
movies["movieId"] = movies["movieId"].astype("int64")
movies.head()

movieId                        title
0      862                    Toy Story
1     8844                      Jumanji
2    15602             Grumpier Old Men
3    31357            Waiting to Exhale
4    11862  Father of the Bride Part II

In [39]:
del metadata

#### We merge the ratings and movies dataframes

In [40]:
df = pd.merge(ratings, movies)

#### There are 5204 movies and 14717 users

In [41]:
df.title.nunique(), df.userId.nunique()

(5204, 14717)

#### We are creating a pivot table, movies are in columns, users are in rows (sparse matrix). It shows which viewer gave which rating to which movie or not.

In [42]:
movie_ratings = df.pivot_table(index = ["userId"], columns = ["title"], values = "rating")
movie_ratings.head()

title   !Women Art Revolution  $5 a Day  'Gator Bait  'R Xmas  \
userId                                                          
1                         NaN       NaN          NaN      NaN   
2                         NaN       NaN          NaN      NaN   
3                         NaN       NaN          NaN      NaN   
4                         NaN       NaN          NaN      NaN   
5                         NaN       NaN          NaN      NaN   

title   'Twas the Night Before Christmas  ...And God Created Woman  \
userId                                                               
1                                    NaN                       NaN   
2                                    NaN                       NaN   
3                                    NaN                       NaN   
4                                    NaN                       NaN   
5                                    NaN                       NaN   

title   ...And the Pursuit of Happiness  00 Schneider - Jagd auf Nihil Baxter  \
userId                                                                          
1                                   NaN                                   NaN   
2                                   NaN                                   NaN   
3                                   NaN                                   NaN   
4                                   NaN                                   NaN   
5                                   NaN                                   NaN   

title   10 Items or Less  10 Things I Hate About You  10,000 BC  \
userId                                                            
1                    NaN                         NaN        NaN   
2                    NaN                         NaN        NaN   
3                    NaN                         NaN        NaN   
4                    NaN                         NaN        NaN   
5                    NaN                         NaN        NaN   

title   11'09''01 - September 11  12 Angry Men  12 Storeys  \
userId                                                       
1                            NaN           NaN         NaN   
2                            NaN           NaN         NaN   
3                            NaN           NaN         NaN   
4                            NaN           NaN         NaN   
5                            NaN           NaN         NaN   

title   12 Wishes of Christmas  127 Hours  13 Tzameti  1408  15 Minutes  \
userId                                                                    
1                          NaN        NaN         NaN   NaN         NaN   
2                          NaN        NaN         NaN   NaN         NaN   
3                          NaN        NaN         NaN   NaN         NaN   
4                          NaN        NaN         NaN   NaN         NaN   
5                          NaN        NaN         NaN   NaN         NaN   

title   16 Blocks  1900  1942: A Love Story  1984  2 Days in Paris  \
userId                                                               
1             NaN   NaN                 NaN   NaN              NaN   
2             NaN   NaN                 NaN   NaN              NaN   
3             NaN   NaN                 NaN   NaN              NaN   
4             NaN   NaN                 NaN   NaN              NaN   
5             NaN   NaN                 NaN   NaN              NaN   

title   2 Fast 2 Furious  20,000 Leagues Under the Sea  2001: A Space Odyssey  \
userId                                                                          
1                    NaN                           NaN                    NaN   
2                    NaN                           NaN                    NaN   
3                    NaN                           NaN                    NaN   
4                    NaN                           NaN                    NaN   
5                    NaN                           NaN                    NaN

#### We will drop the movies which have less than 10 users rated it

In [43]:
movie_ratings = movie_ratings.dropna(thresh = 10, axis = 1).fillna(0)

In [44]:
movie_ratings

title   !Women Art Revolution  'Gator Bait  'Twas the Night Before Christmas  \
userId                                                                         
1                         0.0          0.0                               0.0   
2                         0.0          0.0                               0.0   
3                         0.0          0.0                               0.0   
4                         0.0          0.0                               0.0   
5                         0.0          0.0                               0.0   
...                       ...          ...                               ...   
14995                     0.0          0.0                               0.0   
14996                     0.0          0.0                               0.0   
14997                     0.0          0.0                               0.0   
14998                     0.0          0.0                               0.0   
14999                     0.0          0.0                               0.0   

title   ...And the Pursuit of Happiness  10 Items or Less  \
userId                                                      
1                                   0.0               0.0   
2                                   0.0               0.0   
3                                   0.0               0.0   
4                                   0.0               0.0   
5                                   0.0               0.0   
...                                 ...               ...   
14995                               0.0               0.0   
14996                               0.0               0.0   
14997                               0.0               0.0   
14998                               0.0               0.0   
14999                               0.0               0.0   

title   10 Things I Hate About You  10,000 BC  11'09''01 - September 11  \
userId                                                                    
1                              0.0        0.0                       0.0   
2                              0.0        0.0                       0.0   
3                              0.0        0.0                       0.0   
4                              0.0        0.0                       0.0   
5                              0.0        0.0                       0.0   
...                            ...        ...                       ...   
14995                          0.0        0.0                       0.0   
14996                          0.0        0.0                       0.0   
14997                          0.0        0.0                       0.0   
14998                          0.0        0.0                       0.0   
14999                          0.0        0.0                       0.0   

title   13 Tzameti  1408  15 Minutes  16 Blocks  1900  1984  2 Days in Paris  \
userId                                                                         
1              0.0   0.0         0.0        0.0   0.0   0.0              0.0   
2              0.0   0.0         0.0        0.0   0.0   0.0              0.0   
3              0.0   0.0         0.0        0.0   0.0   0.0              0.0   
4              0.0   0.0         0.0        0.0   0.0   0.0              0.0   
5              0.0   0.0         0.0        0.0   0.0   0.0              0.0   
...            ...   ...         ...        ...   ...   ...              ...   
14995          0.0   0.0         0.0        0.0   0.0   0.0              0.0   
14996          0.0   0.0         0.0        0.0   0.0   0.0              0.0   
14997          0.0   0.0         0.0        0.0   0.0   0.0              0.0   
14998          0.0   0.0         0.0        0.0   0.0   0.0              0.0   
14999          0.0   0.0         0.0        0.0   0.0   0.0              0.0   

title   20,000 Leagues Under the Sea  2001: A Space Odyssey  2010  2046  \
userId                                                                   

#### We are calculating item similarity so we can get recommendation 

In [45]:
item_similarity = movie_ratings.corr(method = "pearson")

#### Get similar movie function

In [46]:
def get_similar_movies(movie_name, user_rating):
    similar_score = item_similarity[movie_name]*(user_rating-2.5)
    similar_score = similar_score.sort_values(ascending = False)
    
    return similar_score

#### We are creating a person who loves science fiction and don't love drama

#### Let's get recommendation

In [47]:
science_fiction_lover = [("Back to the Future Part II", 5), ("Big Fish", 1)]
    
similar_movies = pd.DataFrame()

for movie, rating in science_fiction_lover:
    similar_movies = similar_movies.append(get_similar_movies(movie, rating), ignore_index = True)

similar_movies.head()
similar_movies.sum().sort_values(ascending = False)

Back to the Future Part II    1.966553
Rain Man                      0.891359
Lost in Translation           0.839263
Tough Enough                  0.834341
The Conversation              0.773296
                                ...   
An Unfinished Life           -0.146021
Wedlock                      -0.149409
The Contract                 -0.160847
American Graffiti            -0.168018
Big Fish                     -0.610922
Length: 2591, dtype: float64